tempesta_fw/sched/tfw_sched_ratio.c

```diff
@@ -0,0 +1,467 @@
+			return &sl->epool[i];
+		}
+	}
+
+	return NULL;
+}
+
+/*
+ * Return an entry to the RCU pool.
+ */
+static void
+tfw_sched_rated_put(struct rcu_head *rcup)
+{
+	TfwRatioRated *rated = container_of(rcup, TfwRatioRated, rcu);
+
+	rated->listsz = 0;
+	atomic_set(&rated->free, 1);
+	smp_mb__after_atomic();
+}
+
+/*
+ * Calculate the latest load balancing data if necessary.
+ * This code runs periodically on timer and recalculates each server's
+ * share according to current weight of each server. The calculated data
+ * is used in tfw_sched_ratio_get_srv_conn()
+ *
+ * RCU is used to avoid locks here and in tfw_sched_ratio_get_srv_conn().
+ * When recalculation is required, new data is placed in an available
+ * entry from RCU pool. The new entry is then seamlessly set as the
+ * current entry. The previous entry is made available when all users
+ * of it are gone.
+ * That makes things somewhat non-deterministic. It may happen that no
+ * entry is available at the moment. That's not a big deal. Scheduling
+ * of destination servers will continue to run on current data.
+ * The timer is scheduled to run ASAP and catch a new entry the moment
+ * it gets available.
+ */
+static void
+tfw_sched_ratio_tmfn(unsigned long fndata)
+{
+	TfwSrvGroup *sg = (TfwSrvGroup *)fndata;
+	TfwRatioSrvList *sl = sg->sched_data;
+	uint64_t idx;
+	static uint64_t lastidx = 0;
+	TfwRatioRated *prev_rated, *this_rated;
+	int timeout = TFW_RATIO_TIMEOUT, ratio[sl->srv_n];
+
+	/* No need to do anything if there was no activity. */
+	smp_mb__before_atomic();
+	idx = atomic64_read(&sl->counter);
+	if (unlikely(idx == lastidx))
+		goto rearm;
+
+	/* Just re-arm if no recalculation is required. */
+	if (tfw_sched_ratio_dynamic(sg, ratio))
+		goto rearm;
+	/*
+	 * Get an available entry from the pool. If there's none,
+	 * then re-schedule the timer to run as soon as possible,
+	 * and get an entry the moment it is available.
+	 */
+	this_rated = tfw_sched_rated_get(sl);
+	if (unlikely(!this_rated)) {
+		TFW_DBG3("%s: No available RCU entry\n", __func__);
+		timeout = 1;
+		goto rearm;
+	}
+
+	tfw_sched_fill_rated(sg, ratio, this_rated);
+
+	prev_rated = sl->rated;
+	rcu_assign_pointer(sl->rated, this_rated);
+	call_rcu(&prev_rated->rcu, tfw_sched_rated_put);
+
+	lastidx = idx;
+rearm:
+	smp_mb__before_atomic();
+	if (atomic_read(&sl->rearm))
+		mod_timer(&sl->timer, jiffies + timeout);
+}
+
+/*
+ * Called when a scheduler is set for a server group.
+ * That's when a server group becomes full-functioning.
+ */
+static void
+tfw_sched_ratio_alloc_data(TfwSrvGroup *sg)
+{
+	int i;
+	TfwRatioSrvList *sl;
+
+	BUG_ON(!(sg->flags & ~TFW_SG_F_PSTATS_IDX_MASK));
+
+	sl = kzalloc(sizeof(TfwRatioSrvList), GFP_KERNEL);
+	BUG_ON(!sl);
+
+	for (i = 1; i < TFW_RATIO_EPOOLSZ; ++i)
+		atomic_set(&sl->epool[i].free, 1);
+	rcu_assign_pointer(sl->rated, &sl->epool[0]);
+	sg->sched_data = sl;
+
+	atomic_set(&sl->rearm, 1);
+	setup_timer(&sl->timer, tfw_sched_ratio_tmfn, (unsigned long)sg);
```

Comment:
I think we should setup the timer if there is dynamic scheduling configuration only.